compiler/rustc_passes/src/check_attr.rs

```diff
@@ -2052,6 +2083,90 @@ impl CheckAttrVisitor<'_> {
             errors::Unused { attr_span: attr.span, note },
         );
     }
+
+    fn check_proc_macro(&self, hir_id: HirId, target: Target, kind: ProcMacroKind) {
+        let expected_input_count = match kind {
+            ProcMacroKind::Attribute => 2,
+            ProcMacroKind::Derive | ProcMacroKind::FunctionLike => 1,
+        };
+
+        let expected_signature = match kind {
+            ProcMacroKind::Attribute => "fn(TokenStream, TokenStream) -> TokenStream",
+            ProcMacroKind::Derive | ProcMacroKind::FunctionLike => "fn(TokenStream) -> TokenStream",
+        };
+
+        let tcx = self.tcx;
+        if target == Target::Fn {
+            let Some(tokenstream) = tcx.get_diagnostic_item(sym::TokenStream) else {return};
+            let tokenstream = tcx.type_of(tokenstream);
+
+            let id = hir_id.expect_owner();
+            let hir_sig = tcx.hir().fn_sig_by_hir_id(hir_id).unwrap();
+
+            let sig = tcx.fn_sig(id);
+
+            if sig.abi() != Abi::Rust {
+                tcx.sess
+                    .emit_err(ProcMacroInvalidAbi { span: hir_sig.span, abi: sig.abi().name() });
+                self.abort.set(true);
+            }
+
+            if sig.unsafety() == Unsafety::Unsafe {
+                tcx.sess.emit_err(ProcMacroUnsafe { span: hir_sig.span });
+                self.abort.set(true);
+            }
+
+            let output = sig.output().skip_binder();
+
+            // Typecheck the output
+            if tcx.normalize_erasing_regions(ParamEnv::empty(), output) != tokenstream {
+                tcx.sess.emit_err(ProcMacroTypeError {
+                    span: hir_sig.decl.output.span(),
+                    found: output,
+                    kind,
+                    expected_signature,
+                });
+                self.abort.set(true);
+            }
+
+            // Typecheck "expected_input_count" inputs, emitting
+            // `ProcMacroMissingArguments` if there are not enough.
+            if let Some(args) = sig.inputs().skip_binder().get(0..expected_input_count) {
+                for (arg, input) in args.iter().zip(hir_sig.decl.inputs) {
+                    if tcx.normalize_erasing_regions(ParamEnv::empty(), *arg) != tokenstream {
+                        tcx.sess.emit_err(ProcMacroTypeError {
+                            span: input.span,
+                            found: *arg,
+                            kind,
+                            expected_signature,
+                        });
+                        self.abort.set(true);
+                    }
+                }
+            } else {
+                tcx.sess.emit_err(ProcMacroMissingArguments {
+                    expected_input_count,
+                    span: hir_sig.span,
+                    kind,
+                    expected_signature,
+                });
+                self.abort.set(true);
+            }
```

Comment:
I would invert the `if` statement to avoid the extra nesting of this `for` loop, and combine it with the error below. You can just check that `arguments.len() != expected_input_count`.

Reply:
Alright, I think my approach was better though 🤷‍♂️